core/src/main/java/org/infinispan/container/SegmentedDataContainer.java

```diff
@@ -0,0 +1,168 @@
+package org.infinispan.container;
+
+import java.util.Iterator;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.metadata.Metadata;
+
+/**
+ * Interface describing methods of a data container where operations can be indexed by the segment of the key
+ * stored in the map. This allows for much more efficient iteration when only a subset of segments are required for
+ * a given operation (which is the case very often with Distributed caches).
+ * @author wburns
+ * @since 9.3
+ */
+public interface SegmentedDataContainer<K, V> extends DataContainer<K, V> {
```

Comment:
The new interface added a lot of annoying instanceof checks, how about we add all these methods to `DataContainer` + a method `isSegmented()`?

I was actually thinking of making `SegmentedDataContainer` a super-interface of `DataContainer` with only the methods that we really want to keep post 10.0, and adding default implementations that ignore the segment in `DataContainer`, but I think the only migration path for `AdvancedCache.getDataContainer()` would be to drop it...

Reply:
This needs to be thought out further. Added it to the list of things.